trunner/test_runner.py

```diff
@@ -20,6 +24,69 @@ def resolve_project_path():
     return project_dir
 
 
+def init_logdir():
+    """Inits log directories if it's needed."""
+    if not trunner.ctx.logdir:
+        return
+
+    logdir = trunner.ctx.logdir
+    internal_logdir_names = ("out", "in", "inout")
+    if not os.path.isdir(logdir):
+        os.mkdir(logdir)
+
+    for dirname in internal_logdir_names:
+        if not os.path.isdir(f"{logdir}/{dirname}"):
+            os.mkdir(f"{logdir}/{dirname}")
+        # clear logfiles from the whole test campaign
+        else:
+            with open(f"{trunner.ctx.logdir}/{dirname}/test_campaign.log", "w"):
+                break
+
+
+def set_logfiles(dut: Dut):
+    """Sets dut logfiles associated with the pexpect process if needed."""
+
+    if not trunner.ctx.logdir and not is_github_actions():
+        return
+
+    logfile_r, logfile_w, logfile_a = StringIO(""), StringIO(""), StringIO("")
+    dut.set_logfiles(logfile_r, logfile_w, logfile_a)
+
+
+def format_logs_for_gh(logs_in):
+    """Prepares drop-down list for github actions workflows based on provided raw logs"""
+
+    logs = logs_in
+    # skip esc codes intended to clear window and double cr to avoid printing additional newline
+    logs = logs.replace("\r\r", "\r")
+    logs = logs.replace("\033[2J", "")
+    logs = logs.replace("\033c", "")
+    logs = "::group::show logs\n" + logs + "\n::endgroup::"
+
+    return logs
+
+
+def dump_logfiles(dut: Dut, test):
+    """Dump logfiles to log directory or to gh actions if needed."""
+    if not trunner.ctx.logdir and not is_github_actions():
+        return
+
+    test_basename = test.name.replace("phoenix-rtos-tests/", "")
+    test_basename = test_basename.replace("/", ".")
+
+    for idx, internal_dirname in enumerate(("out", "in", "inout")):
```

Comment:
```suggestion
    for dir, log in zip(("out", "in", "inout"), dut.get_logfiles()):
```
I think this should simplify things! dut.get_logfiles is called once and we don't have to touch any indices.